gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -0,0 +1,449 @@
+      # contains the sample IDs, which is unique per file.
+      self._init_with_header(self._representative_header_lines +
+                             header_lines[-1:])
+    else:
+      self._init_with_header(header_lines)
+
+  def next(self):
+    for text_line in self._text_lines:
+      return self._get_record(text_line)
+    raise StopIteration
+
+  def __iter__(self):
+    return self
+
+  def _init_with_header(self, header_lines):
+    # type: (List[str]) -> None
+    """ This method will be called by _process_header_lines()."""
+    raise NotImplementedError
+
+  def _get_record(self, line):
+    # type: (str) -> Variant
+    """ This method will be called by next(), one line at a time."""
+    raise NotImplementedError
+
+class PyvcfParser(VcfParser):
+  """An Iterator for processing a single VCF file using pyvcf."""
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               file_pattern,  # type: str
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               representative_header_lines=None,  # type:  List[str]
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PyvcfParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+
+  def _init_with_header(self, header_lines):
+    self._header_lines = header_lines
+    try:
+      self._vcf_reader = vcf.Reader(fsock=self._line_generator())
+    except SyntaxError as e:
+      raise ValueError(
+          'Invalid VCF header in %s: %s' % (self._file_name, str(e)))
+
+  def _line_generator(self):
+    header_processed = False
+    if not header_processed and self._header_lines:
+      for header in self._header_lines:
+        self._current_line = header
+        yield self._current_line
+      header_processed = True
+    # Infinit loop; VcfParser.next() rasies StopIteration when EOF is reached.
+    while self._current_line:
+      # PyVCF has explicit str() calls when parsing INFO fields, which fails
+      # with UTF-8 decoded strings. Encode the line back to UTF-8.
+      yield self._current_line.encode('utf-8')
+
+  def _get_record(self, line):
+    # _line_generator will consume this line.
+    self._current_line = line
+    try:
+      record = next(self._vcf_reader)
+      return self._convert_to_variant_record(record, self._vcf_reader.formats)
+    except (LookupError, ValueError) as e:
+      if self._allow_malformed_records:
```

Comment:
I think it's a better design to change the 'contract' of this method (i.e. in the comments) to either return a `Variant` or a `MalformedVcfRecord` (you need to specify this now anyway). That way, you just return the MalformedVcfRecord here. The `next()` method in `VcfParser` then checks whether the type is a Variant or Malformed one and does the right thing depending on allow_malfored_records. That way, for Nucleus, you don't have to re-implement this logic.